lib/pg/instrumentation.rb

```diff
@@ -0,0 +1,185 @@
+          alias_method :prepare_original, :prepare
+          alias_method :exec_prepared_original, :exec_prepared
+
+          def initialize(*args)
+            return initialize_original(*args) if args.empty?
+
+            hash_arg = args.first.is_a?( Hash ) ? args.first : {}
+            db_name = hash_arg.fetch(:dbname, nil)
+            db_user = hash_arg.fetch(:user, nil)
+            host = hash_arg.fetch(:host, nil)
+            port = hash_arg.fetch(:port, nil)
+
+            ::PG::Instrumentation.tags['db.instance'] = db_name if db_name
+            ::PG::Instrumentation.tags['db.user'] = db_user if db_user
+            ::PG::Instrumentation.tags['peer.hostname'] = host if host
+            ::PG::Instrumentation.tags['peer.port'] = port if port
+            ::PG::Instrumentation.tags['peer.address'] = "pg://#{host}:#{port}" if host && port
+
+            operation_name = 'pg.initialize'
+
+            scope = ::PG::Instrumentation.tracer.start_active_span(operation_name, tags: ::PG::Instrumentation.tags)
+
+            initialize_original(*args)
+
+          rescue => e
+            log_error(scope.span, e) if scope
+          ensure
+            scope&.close
+          end
+
+          def async_exec(*args)
+            tags = ::PG::Instrumentation.tags.dup
+
+            sql = args.first.to_s[0, 1024]
+            tags['db.statement'] = sql
+
+            default_op_name = 'pg.query'
+            operation_name = get_operation_name(sql, default_op_name)
+
+            scope = ::PG::Instrumentation.tracer.start_active_span(operation_name, tags: tags)
+
+            async_exec_original(*args)
+
+          rescue => e
+            log_error(scope.span, e) if scope
+          ensure
+            scope&.close
+          end
+
+          def exec(*args)
+            tags = ::PG::Instrumentation.tags.dup
+
+            sql = args.first.to_s[0, 1024]
+            tags['db.statement'] = sql
+
+            default_op_name = 'pg.query'
+            operation_name = get_operation_name(sql, default_op_name)
+            scope = ::PG::Instrumentation.tracer.start_active_span(operation_name, tags: tags)
+
+            exec_original(*args)
+
+          rescue => e
+            log_error(scope.span, e) if scope
+          ensure
+            scope&.close
+          end
+
+          def exec_params(*args)
+            tags = ::PG::Instrumentation.tags.dup
+
+            sql = args.first.to_s[0, 1024]
+            tags['db.statement'] = sql
+
+            default_op_name = 'pg.query'
+            operation_name = get_operation_name(sql, default_op_name)
+
+            scope = ::PG::Instrumentation.tracer.start_active_span(operation_name, tags: tags)
+
+            exec_params_original(*args)
+
+          rescue => e
+            log_error(scope.span, e) if scope
+          ensure
+            scope&.close
+          end
+
+          def prepare(*args)
+            ::PG::Instrumentation.prep_sql = args
```

Comment:
Why at the class level?  I'm not strong in ruby but wouldn't this affect all instances?